// iced-coffee-script/lib/coffee-script/coffee-script.js
// Generated by IcedCoffeeScript 1.4.0a
(function() {
  var EXTENSIONS, Lexer, RESERVED, compile, e, fs, iced, isCoffeeFile, lexer, parser, path, vm, _i, _j, _len, _len1, _ref,
    __hasProp = {}.hasOwnProperty;

  fs = require('fs');

  path = require('path');

  _ref = require('./lexer'), Lexer = _ref.Lexer, RESERVED = _ref.RESERVED;

  parser = require('./parser').parser;

  vm = require('vm');

  iced = require('./iced');

  exports.EXTENSIONS = EXTENSIONS = [".coffee", ".iced"];

  isCoffeeFile = function(file) {
    var e, _i, _len;
    for (_i = 0, _len = EXTENSIONS.length; _i < _len; _i++) {
      e = EXTENSIONS[_i];
      if (path.extname(file) === e) return true;
    }
    return false;
  };

  if (require.extensions) {
    for (_i = 0, _len = EXTENSIONS.length; _i < _len; _i++) {
      e = EXTENSIONS[_i];
      require.extensions[e] = function(module, filename) {
        var content;
        content = compile(fs.readFileSync(filename, 'utf8'), {
          filename: filename
        });
        return module._compile(content, filename);
      };
    }
  } else if (require.registerExtension) {
    for (_j = 0, _len1 = EXTENSIONS.length; _j < _len1; _j++) {
      e = EXTENSIONS[_j];
      require.registerExtension(e, function(content) {
        return compile(content);
      });
    }
  }

  exports.VERSION = '1.4.0a';

  exports.RESERVED = RESERVED;

  exports.helpers = require('./helpers');

  exports.compile = compile = function(code, options) {
    var header, js, merge;
    if (options == null) options = {};
    merge = exports.helpers.merge;
    try {
      js = (iced.transform(parser.parse(lexer.tokenize(code)))).compile(options);
      if (!options.header) return js;
    } catch (err) {
      if (options.filename) {
        err.message = "In " + options.filename + ", " + err.message;
      }
      throw err;
    }
    header = "Generated by IcedCoffeeScript " + this.VERSION;
    return "// " + header + "\n" + js;
  };

  exports.tokens = function(code, options) {
    return lexer.tokenize(code, options);
  };

  exports.nodes = function(source, options) {
    if (typeof source === 'string') {
      return iced.transform(parser.parse(lexer.tokenize(source, options)));
    } else {
      return iced.transform(parser.parse(source));
    }
  };

  exports.run = function(code, options) {
    var mainModule;
    if (options == null) options = {};
    mainModule = require.main;
    mainModule.filename = process.argv[1] = options.filename ? fs.realpathSync(options.filename) : '.';
    mainModule.moduleCache && (mainModule.moduleCache = {});
    mainModule.paths = require('module')._nodeModulePaths(path.dirname(options.filename));
    if ((!isCoffeeFile(mainModule.filename)) || require.extensions) {
      return mainModule._compile(compile(code, options), mainModule.filename);
    } else {
      return mainModule._compile(code, mainModule.filename);
    }
  };

  exports["eval"] = function(code, options) {
    var Module, Script, js, k, o, r, sandbox, v, _k, _len2, _module, _ref1, _ref2, _require;
    if (options == null) options = {};
    if (!(code = code.trim())) return;
    Script = vm.Script;
    if (Script) {
      if (options.sandbox != null) {
        if (options.sandbox instanceof Script.createContext().constructor) {
          sandbox = options.sandbox;
        } else {
          sandbox = Script.createContext();
          _ref1 = options.sandbox;
          for (k in _ref1) {
            if (!__hasProp.call(_ref1, k)) continue;
            v = _ref1[k];
            sandbox[k] = v;
          }
        }
        sandbox.global = sandbox.root = sandbox.GLOBAL = sandbox;
      } else {
        sandbox = global;
      }
      sandbox.__filename = options.filename || 'eval';
      sandbox.__dirname = path.dirname(sandbox.__filename);
      if (!(sandbox !== global || sandbox.module || sandbox.require)) {
        Module = require('module');
        sandbox.module = _module = new Module(options.modulename || 'eval');
        sandbox.require = _require = function(path) {
          return Module._load(path, _module, true);
        };
        _module.filename = sandbox.__filename;
        _ref2 = Object.getOwnPropertyNames(require);
        for (_k = 0, _len2 = _ref2.length; _k < _len2; _k++) {
          r = _ref2[_k];
          if (r !== 'paths') _require[r] = require[r];
        }
        _require.paths = _module.paths = Module._nodeModulePaths(process.cwd());
        _require.resolve = function(request) {
          return Module._resolveFilename(request, _module);
        };
      }
    }
    o = {};
    for (k in options) {
      if (!__hasProp.call(options, k)) continue;
      v = options[k];
      o[k] = v;
    }
    o.bare = true;
    js = compile(code, o);
    if (sandbox === global) {
      return vm.runInThisContext(js);
    } else {
      return vm.runInContext(js, sandbox);
    }
  };

  lexer = new Lexer;

  parser.lexer = {
    lex: function() {
      var tag, _ref1;
      _ref1 = this.tokens[this.pos++] || [''], tag = _ref1[0], this.yytext = _ref1[1], this.yylineno = _ref1[2];
      return tag;
    },
    setInput: function(tokens) {
      this.tokens = tokens;
      return this.pos = 0;
    },
    upcomingInput: function() {
      return "";
    }
  };

  parser.yy = require('./nodes');

  exports.iced = iced.runtime;

}).call(this);
